networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py

```diff
@@ -226,46 +233,111 @@ def _release_vxlan_segment(self, network, host_config, level, segment):
         segment_type = hostgroup.get('segment_type', 'vlan')
         segment_physnet = hostgroup.get('physical_network')
         network_id = network['id']
+        access_id_pool = common.get_set_from_ranges(hostgroup['baremetal_access_vlan_ranges'])
 
-        session = db_api.get_writer_session()
-        segment = session.query(ml2_models.NetworkSegment).filter_by(segmentation_id=segmentation_id,
-                                                                     physical_network=segment_physnet,
-                                                                     network_type=segment_type,
-                                                                     network_id=network_id,
-                                                                     segment_index=level).first()
+        with db_api.exc_to_retry(sa.exc.IntegrityError), session.begin(subtransactions=True):
+            # 1. check if segment exists
+            existing_segments = (session.query(ml2_models.NetworkSegment)
+                                 .filter_by(network_id=network_id, physical_network=segment_physnet,
+                                            segment_index=level, network_type=segment_type)
+                                 .all())
+
+            if existing_segments:
+                if len(existing_segments) > 1:
+                    LOG.error("Multiple segments exists for network %s physical network %s - Segments %s with ids %s",
+                              network_id, segment_physnet, ", ".join(n.id for n in existing_segments),
+                              ", ".join(n.segmentation_id for n in existing_segments))
+                segment = existing_segments[0]
+                if is_access:
+                    # make sure the segment id is from the right pool
+                    if segment.segmentation_id not in access_id_pool:
+                        raise NetworkHasBoundTrunkPorts(network_id=network_id, segment_id=segment.id,
+                                                        segmentation_id=segment.segmentation_id)
+                else:
+                    # make sure the segment id matches, else it's an affinity error
+                    if segment.segmentation_id != segmentation_id:
+                        raise NetworkUsesDifferentTrunkId(network_id=network_id,
+                                                          segmentation_id=segment.segmentation_id)
+                return segment
+
+            # 2. sanity checks
+            if is_access:
+                # for access mode: check that no other network has bound this in host mode
+                host_segments = self.db.get_hosts_on_physnet(ctx, segment_physnet, level=1,
+                                                             with_segment=True, with_segmentation=True)
+                for far_host, far_segment_id, far_segmentation_id in host_segments:
+                    if far_host in hostgroup['hosts'] and far_segmentation_id in access_id_pool:
```

Comment:
Why do we need this condition:
```
far_host in hostgroup['hosts']
```
isn't that implicit?

Reply:
Here I'm getting all hosts that are hosted on a physnet, regardless of the network, meaning all `openstack-baremetal-$project_id` segments. Then, for each segment I check if the host I have found on that segment is part of our current hostgroup and in access mode. If so I raise an exception.

The check is not implicit, as I can have segments for node002-bb095 and node004-bb095, but when node002-bb095 (in this case the host I'm currently trying to bind) is already present on another segment with an access binding I want to raise an exception.